crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,573 @@
+                            );
+
+                            self.chunks.push_back((*new, 0));
+                        }
+
+                        // New chunk at the beginning.
+                        (None, Some(next)) => {
+                            debug_assert!(
+                                matches!(self.chunks.front(), Some((n, _)) if n == next),
+                                "Inserting new chunk at the end: The previous chunk is invalid"
+                            );
+
+                            self.chunks.push_front((*new, 0));
+                        }
+
+                        // New chunk is inserted between 2 chunks.
+                        (Some(previous), Some(next)) => {
+                            let next_chunk_index = self
+                                .chunks
+                                .iter()
+                                .position(|(chunk_identifier, _)| chunk_identifier == next)
+                                // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and
+                                // assuming `Self::chunks` is correctly initialized, it is not
+                                // possible to insert a chunk between two chunks where one does not
+                                // exist. If this predicate fails, it means `LinkedChunk` or
+                                // `Updates` contain a bug.
+                                .expect("Inserting new chunk: The chunk is not found");
+
+                            debug_assert!(
+                                matches!(self.chunks.get(next_chunk_index - 1), Some((p, _)) if p == previous),
+                                "Inserting new chunk: The previous chunk is invalid"
+                            );
+
+                            self.chunks.insert(next_chunk_index, (*new, 0));
+                        }
+
+                        (None, None) => {
+                            unreachable!(
+                                "Inserting new chunk with no previous nor next chunk identifiers \
+                                is impossible"
+                            );
+                        }
+                    }
+                }
+
+                Update::RemoveChunk(expected_chunk_identifier) => {
+                    let chunk_index = self
+                        .chunks
+                        .iter()
+                        .position(|(chunk_identifier, _)| {
+                            chunk_identifier == expected_chunk_identifier
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                        // `Self::chunks` is correctly initialized, it is not possible to remove a
+                        // chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `Updates` contain a bug.
+                        .expect("Removing a chunk: The chunk is not found");
```

Comment:
Why is it better to crash here instead of turning this into a no-op?

Reply:
Doesn't need to be silent, a log line isn't for example. But sure, if it really can't happen.